ir/program/program.rs

```diff
@@ -30,18 +38,234 @@ impl Program {
         &mut self.entry
     }
 
-    pub fn compile(match_: &typeql::query::stage::Match) -> Self {
-        let _entry = FunctionalBlock::from_match(match_);
-        todo!()
+    pub(crate) fn functions(&self) -> &Vec<FunctionIR> {
+        &self.functions
+    }
+
+    pub(crate) fn into_parts(self) -> (FunctionalBlock, Vec<FunctionIR>) {
+        let Self { entry, functions } = self;
+        (entry, functions)
+    }
+
+    // TODO: ProgramDefinitionError?
```

Comment:
Do we want yet another level of errors?
The compilation error can come from the entry pattern.